File: leetCode/medium/3sum/code.py
"""
Given an array nums of n integers, are there elements a, b, c in nums such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

Note:

The solution set must not contain duplicate triplets.

Example:

Given array nums = [-1, 0, 1, 2, -1, -4],

A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]
"""

"""
Test Cases:

[-1, 0, 1] => [ [-1, 0, 1] ]
[-1, 0, -1] => [ ]
[-1, -1, 0, 0, 1, 1] => [ [-1, 0, 1] ]
[-1, -1, 0, 0, 1, 2] => [ [-1, 0, 1], [-1, -1, 2] ]
[-1, 0, 1, -1, 2, 0, 0, 1, -2] => [ [-1, 0, 1], [-1, -1, 2], [1, 1, -2], [0, 0, 0], [2, 0, -2] ]
[ 0, 0, 0, 0 ] => [ [0, 0, 0] ]

Obvious solution: Create sums of every possible triplet, and then remove duplicate triplets

Slightly more clever: Create sums of every unique duo, and use a map to figure out the triplet (if it exists). Still have to remove duplicates...

How to create sums of every unique duo:
[-1, 0, 1] => [ [-1, 0], [-1, 1], [0, 1] ]
           => this would generate the same triplet... three times

Cleverer: We can be slightly greedy about this (search in only one direction, because anything that came before is already part of a triplet)
          => Sort, and remove any exceess that appears more than 3 times
          => Create pairs, and then search to the right
          -- Still O(n^3 + nlogn), but we get rid of removing duplicates

Different Approach: Create two "pools", negatives + positives (+ zeroes are in both pools)
          => For every negative number, see if there are 2 positive numbers that can counter it
          => For every positive number, do the same to the negatives
          => The only duplicates will be the ones with zeros in them. 
Optimization is to take zeroes out of the pools
          => if there are at least 3 zeros, [0,0,0] is part of the solution
          => for any positive number, if there is a negative counterpart, and zero exists, then [+, 0, -] is a solution
Further optimization is to create 4 pools, negatives, positives, sum of 2 negs, sum of 2 pos

Runtime is O(n^2 + n) to create all of the pools. Pools are maps, so searching them is constant runtime
"""

def sum3(narr):
    zeroPool = [ n for n in narr if n == 0 ]
    negPool = { n : 0 for n in narr if n < 0 }
    posPool = { n : 0 for n in narr if n > 0 }
    
    for n in narr:
        if (n < 0):
            negPool[n] += 1
        elif (n > 0):
            posPool[n] += 1
    
    posPoolKeys = posPool.keys()
    posSumPool = {}
    for (aIndex, a) in enumerate(posPoolKeys):
        for b in posPoolKeys[aIndex:]:
            if (a != b or posPool[a] >= 2):
                if (a+b) in posSumPool:
                    posSumPool[a+b].append([a, b])
                else:
                    posSumPool[a+b] = [ [a, b] ]
    negPoolKeys = negPool.keys()
    negSumPool = {}
    for (aIndex, a) in enumerate(negPoolKeys):
        for b in negPoolKeys[aIndex:]:
            if (a != b or negPool[a] >= 2):
                if (a+b) in negSumPool:
                    negSumPool[a+b].append([a, b])
                else:
                    negSumPool[a+b] = [ [a, b] ]
    
    solution = []
    numOfZeroes = len(zeroPool)
    if numOfZeroes >= 3:
        solution.append( [0, 0, 0] )
    
    for neg in negPoolKeys:
        search = -1 * neg
        if search in posSumPool:
            for sumNums in posSumPool[search]:              
                solution.append( [neg] + sumNums )
        if numOfZeroes > 0 and search in posPool:
            solution.append( [neg, 0, search] )

    for pos in posPoolKeys:
        search = -1 * pos
        if search in negSumPool:
            for sumNums in negSumPool[search]:
                solution.append( [pos] + sumNums )
       
    return solution


# Testing
print "Actual: " + str(sum3([1, 0, -1])) + ", Expected: [ [1, 0, -1] ]"
print "Actual: " + str(sum3([0])) + ", Expected: []"
print "Actual: " + str(sum3([1, 0, -1, -1, 2])) + ", Expected: [ [1, 0, -1], [-1, -1, 2] ]"
print "Actual: " + str(sum3([0, 0, 0, 0])) + ", Expected: [ [0, 0, 0] ]"
print "Actual: " + str(sum3([1, 0, -1, 1, 0, -1, 2, -2, 0])) + ", Expected: [ [1, 0, -1], [1, 1, -2], [-1, -1, 2], [0, 0, 0], [2, 0, -2] ]"
print "Actual: " + str(sum3([-4, 2, 2, 3, 1])) + ", Expected: [ [-4, 2, 2], [-4, 3, 1] ]"
print "Actual: " + str(sum3([4, -2, -2, -3, -1])) + ", Expected: [ [4, -2, -2], [4, -3, -1] ]"

print "---------------------------------------------------"

"""
So that worked. 

Except this solution is too expensive space wise, and I seem to have over thought it.

Back to the clever approach, but with one optimization:
When 2Sum'ing the rest of the array, you can use two pointers to only loop through once

This brings solution down to O(n^2 + nlog(n))
"""

def sum3Alt(narr):
    narr.sort()
    
    solution = []
    prevVal = None
    while (len(narr) >= 3):
        val = narr[0]
        narr = narr[1:]
        
        if val == prevVal: 
            continue
        
        leftIndex = 0
        rightIndex = len(narr) - 1
        
        while (leftIndex < rightIndex):
            sumValue = val + narr[leftIndex] + narr[rightIndex]
            if (sumValue == 0):
                solution.append([val, narr[leftIndex], narr[rightIndex]])
                # move left and right indexes to new numbers
                while (leftIndex < rightIndex and narr[leftIndex] == narr[leftIndex + 1]):
                    leftIndex += 1
                while (leftIndex < rightIndex and narr[rightIndex] == narr[rightIndex - 1]):
                    rightIndex -= 1
                leftIndex += 1
                rightIndex -= 1
            elif (sumValue < 0):
                leftIndex += 1
            elif (sumValue > 0):
                rightIndex -= 1

        prevVal = val

    return solution


print "Actual: " + str(sum3Alt([1, 0, -1])) + ", Expected: [ [1, 0, -1] ]"
print "Actual: " + str(sum3Alt([0])) + ", Expected: []"
print "Actual: " + str(sum3Alt([1, 0, -1, -1, 2])) + ", Expected: [ [1, 0, -1], [-1, -1, 2] ]"
print "Actual: " + str(sum3Alt([0, 0, 0, 0])) + ", Expected: [ [0, 0, 0] ]"
print "Actual: " + str(sum3Alt([1, 0, -1, 1, 0, -1, 2, -2, 0])) + ", Expected: [ [1, 0, -1], [1, 1, -2], [-1, -1, 2], [0, 0, 0], [2, 0, -2] ]"
print "Actual: " + str(sum3Alt([-4, 2, 2, 3, 1])) + ", Expected: [ [-4, 2, 2], [-4, 3, 1] ]"
print "Actual: " + str(sum3Alt([4, -2, -2, -3, -1])) + ", Expected: [ [4, -2, -2], [4, -3, -1] ]"

File: leetCode/medium/3sum-closest/code.js
/*
Given an array nums of n integers and an integer target, find three integers in nums such that the sum is closest to target. 
Return the sum of the three integers. You may assume that each input would have exactly one solution.

Example:

Given array nums = [-1, 2, 1, -4], and target = 1.

The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
*/

/*
Test cases
[0, 0, 0] n => 0
[1, 2, 3, 4] 4 => 6
[1, -1, 0, 2] 3 => 3
[0, 3, 5, 9] 10 => 8

Theory:
Obvious solution: generate all combinations of numbers, and find the closest sum. Runtime would be O(n^3), maybe reduced with some optimizations
Optimization to obvious solution:                   
                   => Sort the array. For every num in the array, we can "include" it as one of the possible numbers
                   => Create left + right pointers at the ends of the array. Move pointers closer to the "included" number, depending on dist from target
                   => This would create a runtime of O(n^2 + nlogn)

Slightly more clever: Since we only care about the sum, we don't really need the combination of numbers; just playing with values should be enough
                   => Use a similar pointer system and sorted arr, but this time start the "center" pointer right by the left one.
                   => Store the "closest" value so far
                   => Bounce the "center" pointer between the left and right pointer?
                   
*/

function sum3Closest(nums, target) {
    if (nums.length < 3) return nums.reduce((a, b) => a + b, 0);

    nums.sort((a,b) => a-b);
        
    let closest = nums[0] + nums[1] + nums[2];
    let tryIndex = 0;
    while (tryIndex < nums.length - 2) {
        let leftIndex = tryIndex + 1;
        let rightIndex = nums.length - 1;
        while (leftIndex < rightIndex) {
            let sum = nums[tryIndex] + nums[leftIndex] + nums[rightIndex];
            if (sum === target) return target;

            let closeness = sum - target;
            
            if (Math.abs(closeness) < Math.abs(target - closest)) {
                closest = sum;
            }
            
            if (closeness > 0) {
                rightIndex -= 1;
            } else if (closeness < 0) {
                leftIndex += 1;
            }
        }
        tryIndex += 1;
    }

    return closest;
}


console.log(`Actual: ${sum3Closest([0,0,0], 3)}, Expected: 0`);
console.log(`Actual: ${sum3Closest([1,2,3], 3)}, Expected: 6`);
console.log(`Actual: ${sum3Closest([1,0,0,1], 3)}, Expected: 2`);
console.log(`Actual: ${sum3Closest([4,1,3,-8,8], 0)}, Expected: 1`);
console.log(`Actual: ${sum3Closest([0,3,5,9], 10)}, Expected: 8`);
console.log(`Actual: ${sum3Closest([1,2,4,8,16,32,64,128], 82)}, Expected: 82`);
